## Importing the libraries

In [84]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import ydata_profiling as yp
from ydata_profiling import ProfileReport
import warnings

### Objectives:
- Our complete analysis depend on the loan amount.
- We will build model for this purpose.
- And also analyse the data on its basis.

## Set some options

In [20]:
pd.set_option('display.max_rows',None)
pd.set_option('display.max_columns',None)
warnings.filterwarnings('ignore')

## Loading the dataset

In [82]:
df = pd.read_csv('C:/Bank_loan/mwb2014.csv')

In [86]:
# Generate the profile report
profile = ProfileReport(df, title="Bank Loan Dataset Profiling Report", explorative=True)
# Save the report to an HTML file
profile.to_file("bank_loan_profiling_report.html")

Export report to file: 100%|██████████| 1/1 [00:00<00:00, 29.48it/s]


In [66]:
df.shape

(2501, 24)

In [65]:
df.columns

Index(['First Name', 'Last Name', 'City', 'State', 'Zip', 'Loan Amount',
       'Estimated Value', 'Loan Purpose', 'Qualification FICO',
       'Is Primary Wage Earner', 'Income Type', 'Asset Docs', 'Occupancy',
       'Unit Type', 'Loan Type', 'Lender Name', 'Created Date', 'Time', 'Fix',
       'ARM', 'CLTV', 'Loan Number', 'Loan Officer Name', 'LoanInMonth'],
      dtype='object')

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2501 entries, 0 to 2500
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   First Name              2501 non-null   object 
 1   Last Name               2501 non-null   object 
 2   City                    2501 non-null   object 
 3   State                   2501 non-null   object 
 4   Zip                     2501 non-null   int64  
 5   Loan Amount             2501 non-null   float64
 6   Estimated Value         2501 non-null   int64  
 7   Loan Purpose            2501 non-null   object 
 8   Qualification FICO      2501 non-null   int64  
 9   Is Primary Wage Earner  2501 non-null   bool   
 10  Income Type             2501 non-null   object 
 11  Asset Docs              2501 non-null   object 
 12  Occupancy               1846 non-null   object 
 13  Unit Type               2501 non-null   object 
 14  Loan Type               2501 non-null   

### number of unique values

In [87]:
df.nunique()

First Name                1743
Last Name                 1378
City                       677
State                        6
Zip                        484
Loan Amount               1162
Estimated Value            618
Loan Purpose                 7
Qualification FICO         227
Is Primary Wage Earner       2
Income Type                  4
Asset Docs                   3
Occupancy                    3
Unit Type                   12
Loan Type                    6
Lender Name                 13
Created Date               976
Time                       879
Fix                          2
ARM                          2
CLTV                       247
Loan Number               2476
Loan Officer Name           44
LoanInMonth                 78
dtype: int64

In [29]:
df.State.unique()

array(['NJ', 'NY', 'FL', 'NC', 'PA', 'CT'], dtype=object)

In [31]:
df['Loan Purpose'].unique()

array(['Purchase', 'Cash-Out Refi', 'No Cash Out Refi', 'Reverse',
       'Other', 'cash-Out Refi', 'Contruction- Perm'], dtype=object)

In [33]:
df['Income Type'].unique()

array(['Full', 'Stated', 'NoRatio', 'NoEmployment'], dtype=object)

In [34]:
df['Asset Docs'].unique()

array(['Full', 'Stated', 'NoAsset'], dtype=object)

In [35]:
df['Occupancy'].unique()

array([nan, 'Owner Occupied', 'Investor', 'Second Home'], dtype=object)

In [36]:
df['Loan Type'].unique()

array(['Residential', 'FHA', 'Commercial', 'Conventional', 'Other', 'VA'],
      dtype=object)

In [37]:
df['Lender Name'].unique()

array(['MWBI', 'HB', 'HCS', 'WFB', 'FSTR', 'CB', 'MTB', 'NP', 'PDL',
       'HVER', 'JCB', 'MFSB', 'FSB'], dtype=object)

In [80]:
df['Unit Type'].unique()

array(['TwoFamily', 'Condo', 'OneFamily', 'Coop', 'FourFamily',
       'ThreeFamily', 'PUD', 'MixedUse', 'MultiFamily', 'Warehouse',
       'Industrial', 'Land'], dtype=object)

In [38]:
df['LoanInMonth'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68,
       69, 70, 71, 72, 73, 74, 75, 76, 77, 78], dtype=int64)

In [23]:
df.sample(8)

First Name Last Name            City State    Zip  Loan Amount  \
221          83ck  1rgyr4s          Astoria    NY  11102    625000.00   
1832  Chr3st49h2r       622  Palisades Park    NJ   7650    315000.00   
236          1671  C1S38G16     Jersey City    NJ   7304    155325.00   
327          1T3F  74H1771D        Bethpage    NY  11714    361517.00   
1004       Q3F28G       638   Valley Stream    NY  11580    285000.00   
764       R16st48     1d17s        Westbury    NY  11590    242673.75   
1795        J4rg2   J1ch2r4      Bridgeport    CT   6606    300366.00   
1109         1GH1      1Z17       Brooklyn     NY  11203    594474.38   

      Estimated Value   Loan Purpose  Qualification FICO  \
221            900000  Cash-Out Refi                 700   
1832           420000       Purchase                 704   
236            163500  Cash-Out Refi                 771   
327            374000       Purchase                 752   
1004           380000       Purchase                 717   
764            265000       Purchase                 672   
1795           328000       Purchase                 681   
1109           615000       Purchase                 649   

      Is Primary Wage Earner Income Type Asset Docs       Occupancy  \
221                     True        Full       Full             NaN   
1832                   False        Full       Full  Owner Occupied   
236                     True        Full       Full             NaN   
327                     True        Full       Full             NaN   
1004                   False        Full       Full  Owner Occupied   
764                     True        Full       Full  Owner Occupied   
1795                   False        Full       Full  Owner Occupied   
1109                   False        Full       Full  Owner Occupied   

        Unit Type     Loan Type Lender Name Created Date       Time    Fix  \
221   ThreeFamily   Residential        MWBI    2/12/2015    9:05 AM  False   
1832        Condo  Conventional        MWBI    9/21/2017    4:04 PM   True   
236     OneFamily   Residential        MWBI    2/25/2015   10:25 AM   True   
327     OneFamily           FHA        MWBI    4/13/2015    2:26 PM   True   
1004    OneFamily  Conventional        MWBI    5/10/2016   10:15 PM   True   
764     OneFamily           FHA        MWBI   12/11/2015    9:19 PM   True   
1795  ThreeFamily           FHA        MWBI    8/23/2017    5:40 PM   True   
1109    TwoFamily           FHA         WFB    7/18/2016    7:08 PM  False   

        ARM   CLTV    Loan Number Loan Officer Name  LoanInMonth  
221    True  45.00   20220299998A      J377y V1tt2s           29  
1832  False  75.00     4094023CJA     J2ss3c1 Ch58g           23  
236   False  74.56    202404499JA     V3v318 D2g531           44  
327   False  95.00  G20430259998A   7189r22t Bh1s38           25  
1004  False  75.00   30520199998A  73ch126 61g45d3s           19  
764   False  90.00   G2122059998A   2dw38 81v1rr2t2           18  
1795  False  90.00   G408403699CA  73ch126 61g45d3s           36  
1109   True  95.00   G3074027B98A     74h172d W163d           27

### Dupicate values in dataset

In [88]:
df.duplicated().sum()

9

> Observation:  There are nine records whcich have duplicated values

### Removing the duplicates

In [89]:
df.drop_duplicates(inplace=True)

In [90]:
df.duplicated().sum()

0

In [91]:
df.shape

(2492, 24)

### Checking the null values

In [26]:
df.isnull().sum()

First Name                  0
Last Name                   0
City                        0
State                       0
Zip                         0
Loan Amount                 0
Estimated Value             0
Loan Purpose                0
Qualification FICO          0
Is Primary Wage Earner      0
Income Type                 0
Asset Docs                  0
Occupancy                 655
Unit Type                   0
Loan Type                   0
Lender Name                 0
Created Date                0
Time                        0
Fix                         0
ARM                         0
CLTV                        0
Loan Number                 0
Loan Officer Name           0
LoanInMonth                 0
dtype: int64

In [27]:
np.round((df.isnull().sum()/len(df)) * 100,2)

First Name                 0.00
Last Name                  0.00
City                       0.00
State                      0.00
Zip                        0.00
Loan Amount                0.00
Estimated Value            0.00
Loan Purpose               0.00
Qualification FICO         0.00
Is Primary Wage Earner     0.00
Income Type                0.00
Asset Docs                 0.00
Occupancy                 26.19
Unit Type                  0.00
Loan Type                  0.00
Lender Name                0.00
Created Date               0.00
Time                       0.00
Fix                        0.00
ARM                        0.00
CLTV                       0.00
Loan Number                0.00
Loan Officer Name          0.00
LoanInMonth                0.00
dtype: float64

> Observation : There is only one feature `Occupancy` that have missing values 26.19%.

### Imputing the null values

In [92]:
df['Occupancy'].fillna(df['Occupancy'].mode()[0],inplace=True)

### Checking the null values after imputing

In [93]:
df.isnull().sum()

First Name                0
Last Name                 0
City                      0
State                     0
Zip                       0
Loan Amount               0
Estimated Value           0
Loan Purpose              0
Qualification FICO        0
Is Primary Wage Earner    0
Income Type               0
Asset Docs                0
Occupancy                 0
Unit Type                 0
Loan Type                 0
Lender Name               0
Created Date              0
Time                      0
Fix                       0
ARM                       0
CLTV                      0
Loan Number               0
Loan Officer Name         0
LoanInMonth               0
dtype: int64

> Now our data is cleaned from missing values.

### Histogram of Loan Amount

In [78]:

fig=px.histogram(df,x='Loan Amount',title='Histogram of Loan Amount')
fig.show()

### Separate the categorical and numerical features

### Bar plot of `City` on the base of `Loan Amount`

In [96]:
# Aggregate data: compute the average loan amount per city
agg_df = df.groupby('City', as_index=False)['Loan Amount'].mean()

# Create a bar plot to show the average Loan Amount by City
fig = px.bar(agg_df, x='City', y='Loan Amount',
            title='Average Loan Amounts by City',
            labels={'Loan Amount': 'Average Loan Amount ($)', 'City': 'City'})

# Update layout to improve readability (optional)
fig.update_layout(
    xaxis=dict(
        tickangle=-90,  # Rotate x-axis labels for better visibility
        showticklabels=True
    ),
    height=600,  # Set the height of the plot
    showlegend=False
)
fig.show()

### Bar plot of `Created Date` on the base of `Loan Amount`

In [99]:
agg_df = df.groupby('Created Date', as_index=False)['Loan Amount'].mean()

# Create a bar plot to show the average Loan Amount by City
fig = px.bar(agg_df, x='Created Date', y='Loan Amount',
            title='Average Loan Amounts by City',
            labels={'Loan Amount': 'Average Loan Amount ($)', 'Created Date': 'Created Date'})

# Update layout to improve readability (optional)
fig.update_layout(
    xaxis=dict(
        tickangle=-90,  # Rotate x-axis labels for better visibility
        showticklabels=True
    ),
    height=600,  # Set the height of the plot
    showlegend=False
)
fig.show()

### Bar plot of `Loan Officer Name` on the base of `Loan Amount`

In [125]:
agg_df = df.groupby('Loan Officer Name', as_index=False)['Loan Amount'].sum()

# Create a bar plot to show the average Loan Amount by City
fig = px.bar(agg_df, x='Loan Officer Name', y='Loan Amount',
            title='Average Loan Amounts by City',
            labels={'Loan Amount': 'Average Loan Amount ($)', 'Loan Officer Name': 'Loan Officer Name'})

# Update layout to improve readability (optional)
fig.update_layout(
    xaxis=dict(
        tickangle=-90,  # Rotate x-axis labels for better visibility
        showticklabels=True
    ),
    height=600,  # Set the height of the plot
    showlegend=False
)
fig.show()

### Bar plot of `Time` on the base of `Loan Amount`

In [124]:
agg_df = df.groupby('Time', as_index=False)['Loan Amount'].sum()

# Create a bar plot to show the average Loan Amount by City
fig = px.bar(agg_df, x='Time', y='Loan Amount',
            title='Average Loan Amounts by City',
            labels={'Loan Amount': 'Average Loan Amount ($)', 'Time': 'Time'})

# Update layout to improve readability (optional)
fig.update_layout(
    xaxis=dict(
        tickangle=-90,  # Rotate x-axis labels for better visibility
        showticklabels=True
    ),
    height=600,  # Set the height of the plot
    showlegend=False
)
fig.show()

In [123]:
# List of categorical columns
cat_col = ['State', 'Loan Purpose', 'Is Primary Wage Earner', 'Income Type', 'Asset Docs', 
        'Occupancy', 'Unit Type', 'Loan Type', 'Lender Name', 'Fix', 'ARM']

# Generate and display bar plots for each categorical column
for col in cat_col:
    # Aggregate data: compute the average loan amount per category
    agg_df = df.groupby(col, as_index=False)['Loan Amount'].mean()
    
    # Create a bar plot to show the average Loan Amount by category
    fig = px.bar(agg_df, x=col, y='Loan Amount',
                title=f'Average Loan Amounts by {col}',
                labels={'Loan Amount': 'Average Loan Amount ($)', col: col})
    
    # Show the plot
    fig.show()

In [120]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2492 entries, 0 to 2500
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   First Name              2492 non-null   object 
 1   Last Name               2492 non-null   object 
 2   City                    2492 non-null   object 
 3   State                   2492 non-null   object 
 4   Zip                     2492 non-null   int64  
 5   Loan Amount             2492 non-null   float64
 6   Estimated Value         2492 non-null   int64  
 7   Loan Purpose            2492 non-null   object 
 8   Qualification FICO      2492 non-null   int64  
 9   Is Primary Wage Earner  2492 non-null   bool   
 10  Income Type             2492 non-null   object 
 11  Asset Docs              2492 non-null   object 
 12  Occupancy               2492 non-null   object 
 13  Unit Type               2492 non-null   object 
 14  Loan Type               2492 non-null   objec

In [121]:
num_col = ['Zip','Estimated Value','Qualification FICO','CLTV','Loan Number','LoanInMonth']